cpp/daal/src/algorithms/service_kernel_math.h

```diff
@@ -270,6 +282,48 @@ class EuclideanDistances : public PairwiseDistances<FPType, cpu>
     TArray<FPType, cpu> normBufferB;
 };
 
+// compute: A*B' / (sum(A^2, 2) * sum(B^2, 2))
+template <typename FPType, CpuType cpu>
+class CosineDistances : public EuclideanDistances<FPType, cpu>
+{
+public:
+    using super = EuclideanDistances<FPType, cpu>;
+
+    CosineDistances(const NumericTable & a, const NumericTable & b) : super(a, b, true) {}
+
+    virtual ~CosineDistances() {}
+
+    virtual PairwiseDistanceType getType() { return PairwiseDistanceType::cosine; }
+
+    // output:  Row-major matrix of size { aSize x bSize }
+    services::Status computeBatch(const FPType * const a, const FPType * const b, size_t aOffset, size_t aSize, size_t bOffset, size_t bSize,
+                                  FPType * const res) override
+    {
+        const size_t nRowsA = aSize;
+        const size_t nColsA = super::_a.getNumberOfColumns();
+        const size_t nRowsB = bSize;
+
+        const size_t nRowsC = nRowsA;
+        const size_t nColsC = nRowsB;
+
+        super::computeABt(a, b, nRowsA, nColsA, nRowsB, res);
+
+        const FPType * const aa = super::normBufferA.get() + aOffset;
+        const FPType * const bb = (&(super::_a) == &(super::_b)) ? super::normBufferA.get() + bOffset : super::normBufferB.get() + bOffset;
+
+        PRAGMA_IVDEP
+        PRAGMA_VECTOR_ALWAYS
+        for (size_t i = 0; i < nRowsC; i++)
+        {
+            for (size_t j = 0; j < nColsC; j++)
+            {
+                res[i * nColsC + j] = FPType(1) - (res[i * nColsC + j] / Math<FPType, cpu>::sSqrt(aa[i] * bb[j]));
```

Comment:
sSqrt can be to slow, better to try vSqrt somehow

Reply:
For example use TLS for tmp sqrt results

Reply:
@SmirnovEgorRu 
and division too. I want to move into init and division and root. but later. will be:
```suggestion
                res[i * nColsC + j] = FPType(1) - (res[i * nColsC + j] * aa[i] * bb[j]));
```